internal/clients/state/v3/state.go

```diff
@@ -108,7 +116,11 @@ func NewClient(ctx context.Context, pbClient pb.PluginClient, tableName string)
 			keys := record.Columns()[0].(*array.String)
 			values := record.Columns()[1].(*array.String)
 			for i := 0; i < keys.Len(); i++ {
-				c.mem[keys.Value(i)] = values.Value(i)
+				k, v := keys.Value(i), values.Value(i)
+				curVal, ok := c.mem[k]
+				if !ok || len(curVal) != len(v) || strings.Compare(curVal, v) == -1 { // set only if not exists, of different length, or greater lexicographically
```

Comment:
Write only if it's a new key, or the length of value differs (because we give up sorting at this stage)
if both conditions don't suit, only write if we're writing a "larger" value with the same length (`value2` will overwrite `value1`, but `value0` won't)